server/src/downloaders/photo-downloader.ts

```diff
@@ -1,3 +1,19 @@
+import http from "http";
+import https from "https";
+import Stream from "stream";
+import fs from "fs";
+
 export const downloadPhotoWithCache = ({url, cacheDir}: {url: string; cacheDir: string}) => {
```

Comment:
Oh I just realized `downloadPhotoWithCache` and `downloadSoundWithCache` are required to return the filename of the downloaded file (Since the caller will have no idea what the final filename is).

The filename can be sha1 hash of the response data + the extension


This function can get the hash of a given arrayBuffer (Gotta run  `npm install sha1uint8array` in the server)

```
import {createHash} from 'sha1uint8array';

const getHashForResponseArrayBuffer = (arrayBuffer: ArrayBuffer) =>  createHash().update(arrayBuffer).digest("hex");
```

Also, this function can resolve the extension of the downloaded file 

```
export function resolveExtFromResponseHeaders(response: Response): string | undefined {
    const map: { [key: string]: string } = {
        'image/jpeg': 'jpeg',
        'image/png': 'png',
        'image/gif': 'gif',
    };
    return map[response.headers.get('content-type') as string] || undefined;
}

```


So to write a response to a file with a filename from the hash, and the extension would look like this

```
const response = await fetch(url);
const responseArrayBuffer = await response.arrayBuffer()
const fileExtension = resolveExtFromResponseHeaders(response);
const fileHash  = getHashForResponseArrayBuffer(responseArrayBuffer);
const filePath = join(cacheDir, `${fileHash}.${fileExtension}`);
await fs.writeFile(filePath);
```